vxquery-core/src/main/java/org/apache/vxquery/compiler/rewriter/rules/PushValueIntoDatascanRule.java

```diff
@@ -0,0 +1,148 @@
+import org.apache.vxquery.metadata.VXQueryIndexingDataSource;
+import org.apache.vxquery.metadata.VXQueryMetadataProvider;
+import org.apache.vxquery.functions.BuiltinOperators;
+
+import org.apache.commons.lang3.ArrayUtils;
+
+/**
+ * The rule searches for two assign operators immediately following a data scan
+ * operator.
+ *
+ * <pre>
+ * Before
+ *
+ *   plan__parent
+ *   ASSIGN( $v3 : value( $v1, $v2 ) )
+ *   ASSIGN( $v2 : constant )
+ *   DATASCAN( $source : $v1 )
+ *   plan__child
+ *
+ *   Where $v1 and $v2 is not used in plan__parent.
+ *
+ * After
+ *
+ *   plan__parent
+ *   ASSIGN( $v3 : $v1 )
+ *   ASSIGN( $v2 : constant )
+ *   DATASCAN( $source : $v1 )
+ *   plan__child
+ *
+ *   $source is encoded with the value parameters.
+ * </pre>
+ */
+
+public class PushValueIntoDatascanRule extends AbstractUsedVariablesProcessingRule {
+	StaticContext dCtx = null;
+
+	@Override
+	protected boolean processOperator(Mutable<ILogicalOperator> opRef, IOptimizationContext context)
+			throws AlgebricksException {
+		if (dCtx == null) {
+			VXQueryOptimizationContext vxqueryCtx = (VXQueryOptimizationContext) context;
+			dCtx = ((VXQueryMetadataProvider) vxqueryCtx.getMetadataProvider()).getStaticContext();
+		}
+		AbstractLogicalOperator op1 = (AbstractLogicalOperator) opRef.getValue();
+		if (op1.getOperatorTag() != LogicalOperatorTag.ASSIGN) {
+			return false;
+		}
+		AssignOperator assign = (AssignOperator) op1;
+
+		AbstractLogicalOperator op4 = (AbstractLogicalOperator) assign.getInputs().get(0).getValue();
+		if (op4.getOperatorTag() != LogicalOperatorTag.DATASOURCESCAN) {
+			return false;
+		}
+		DataSourceScanOperator datascan = (DataSourceScanOperator) op4;
+
+		if (!usedVariables.contains(datascan.getVariables())) {
+			VXQueryCollectionDataSource ds = null;
+			VXQueryIndexingDataSource ids = null;
+
+			// Find all value functions.
+			try {
+				ids = (VXQueryIndexingDataSource) datascan.getDataSource();
+			} catch (ClassCastException e) {
+				ds = (VXQueryCollectionDataSource) datascan.getDataSource();
+			}
+
+			if (!updateDataSource(ds, assign.getExpressions().get(0))) {
+				return false;
+			}
+			// Replace assign with noop assign. Keeps variable chain.
+			Mutable<ILogicalExpression> varExp = ExpressionToolbox
+					.findVariableExpression(assign.getExpressions().get(0), datascan.getVariables().get(0));
+			AssignOperator noOp = new AssignOperator(assign.getVariables().get(0), varExp);
+			noOp.getInputs().addAll(assign.getInputs());
+			opRef.setValue(noOp);
+			return true;
+		}
+		return false;
+	}
+
+	private boolean updateDataSource(VXQueryCollectionDataSource ds, Mutable<ILogicalExpression> expression) {
+		boolean added = false;
+		ILogicalExpression comparison = null;
+		List<Mutable<ILogicalExpression>> finds = new ArrayList<Mutable<ILogicalExpression>>();
+		ExpressionToolbox.findAllFunctionExpressions(expression, BuiltinOperators.VALUE.getFunctionIdentifier(), finds);
+		if (finds.size() > 0) {
```

Comment:
Can you explain what your doing here? What do the following variable do: comparison, b, and one?